src/Radicle/Internal/MachineBackend/Ipfs.hs

```diff
@@ -0,0 +1,318 @@
+
+
+ipfsMachineBackend :: MonadIO m => MachineBackend MachineEntryIndex m
+ipfsMachineBackend =
+    MachineBackend
+        { machineType = "ipfs"
+        , machineUpdate =
+            \id values -> do
+                res <- liftIO $ tryAny $ sendIpfs id values
+                pure $ first (toS . displayException) res
+        , machineGetLog =
+            \id maybeFrom -> do
+                res <- liftIO $ tryAny $ receiveIpfs id maybeFrom
+                pure $ first (toS . displayException) res
+        }
+
+sendIpfs :: Text -> Seq Value -> IO MachineEntryIndex
+sendIpfs ipnsId values = do
+    IpfsNameResolveResponse cid <- nameResolve ipnsId
+    IpfsDagPutResponse newEntryCid <- dagPut $ MachineEntry (toList values) cid
+    namePublish ipnsId $ IpfsAddressIpfs newEntryCid
+    pure $ MachineEntryIndex newEntryCid
+
+receiveIpfs :: IpnsId -> Maybe MachineEntryIndex -> IO (MachineEntryIndex, [Value])
+receiveIpfs ipnsId maybeFrom = do
+    let MachineEntryIndex fromCid = fromMaybe (MachineEntryIndex emtpyMachineCid) maybeFrom
+    IpfsNameResolveResponse cid <- nameResolve ipnsId
+    blocks <- getBlocks cid fromCid
+    pure $ (MachineEntryIndex cid, blocks)
+  where
+    getBlocks :: CID -> CID -> IO [Value]
+    getBlocks cid fromCid = do
+        if cid == fromCid || cid == emtpyMachineCid
+        then pure []
+        else do
+            entry <- dagGet (IpfsAddressIpfs cid)
+            rest <- getBlocks (entryPrevious entry) fromCid
+            pure $ rest <> entryExpressions entry
+
+-- | If a machine points to this ID then its log is considered empty.
+-- The first entry in a machine log also points to this entry.
+--
+-- This is the CID produced by the document @{"radicle": true}@.
+emtpyMachineCid :: CID
+emtpyMachineCid =
+    case cidFromText "zdpuAyyGtvC37aeZid2zh7LAGKCbFTn9MzdqoPpbNQm3BCwWT" of
+        Left e    -> panic $ toS e
+        Right cid -> cid
+
+-- | A node in the linked list of expressions of a machine.
+data MachineEntry = MachineEntry
+    { entryExpressions :: [Value]
+    , entryPrevious    :: CID
+    } deriving (Eq, Show, Read, Generic)
+
+instance FromJSON MachineEntry where
+    parseJSON = Aeson.withObject "MachineEntry" $ \o -> do
+        expressionCode <- o .: "expression"
+        let src = "[ipfs]"
+        entryExpressions <-
+            case parseValues src expressionCode of
+                Left err  -> fail $ "failed to parse Radicle expression: " <> show err
+                Right v -> pure v
+        entryPrevious <- parseIpldLink =<< o .: "previous"
+        pure MachineEntry {..}
+
+instance ToJSON MachineEntry where
+    toJSON MachineEntry{..} =
+        let code = T.intercalate "\n" $ map renderCompactPretty entryExpressions
+        in Aeson.object
+            [ "expression" .= code
+            , "previous" .= ipldLink entryPrevious
+            ]
+
+-- | Given a CID @"abc...def"@ it returns a IPLD link JSON object
+-- @{"/": "abc...def"}@.
+ipldLink :: CID -> Aeson.Value
```

Comment:
This would be nice to have in [oscoin/ipfs](https://github.com/oscoin/ipfs) somewhere

Reply:
Yes indeed. There is a couple of things we should backport, like the response types.